conjure_oxide/tests/integration/basic/sum/01-deeply-nested/input-expected-rule-trace.json

```diff
@@ -1,4 +1,27 @@
-{"message":"Rule applicable: normalise_associative_commutative ([(\"Base\", 8400)]), to expression: Sum([a, Sum([b, Sum([c, d])]), e]), resulting in: Sum([a, b, c, d, e])","target":"rule_engine"}
-{"message":"Rule applicable: sum_eq_to_sumeq ([(\"Minion\", 4400)]), to expression: (Sum([a, b, c, d, e]) = 5), resulting in: SumEq([a, b, c, d, e], 5)","target":"rule_engine"}
-{"message":"Rule applicable: sumeq_to_minion ([(\"Minion\", 4400)]), to expression: SumEq([a, b, c, d, e], 5), resulting in: And([SumGeq([a, b, c, d, e], 5), SumLeq([a, b, c, d, e], 5)])","target":"rule_engine"}
-{"count":3,"message":" Number of rules applied"}
+{"message":"Rule applicable: sum_eq_to_sumeq ([(\"Minion\", 4400)]), to expression: (Sum([a, Sum([b, Sum([c, d])]), e]) = 5), resulting in: SumEq([a, Sum([b, Sum([c, d])]), e], 5)","target":"rule_engine"}
```

Comment:
Is this related to do with the rewriter flaw where we always do bigger expressions regardless of priority?

Reply:
could be! if so, good motivating example for @lixitrixi @YehorBoiar

Reply:
If so, we will need to invent  some extra checks to put on this rule.....

Reply:
@ozgurakgun as a stop-gap, I could implement a completely unoptimised naive rewriter where I just iterate over all expressions (with Uniplate) and apply the highest priority rule? I think our models are still small enough where this would somewhat work...

Reply:
I have no objections to that.

Reply:
@ozgurakgun to double check, the only important thing about the new rewriter system that would affect correctness is that it applies the highest priority rule regardless of locality?

Reply:
also, if there two rules at the same level that apply to X and Y, and X is contained in Y, then we choose the one that applies on Y.

but this should come out for free with the universe order of uniplate unless I am misremembering things. assuming you choose the first applicable one out of the highest priority options.